ci-operator-internals: add testing page (#235)
openshift/ci-docs: content/en/docs/architecture/ci-operator-internals/testing.md

image, since it is a regular, statically-linked Go binary:

{{< highlight bash >}}
$ podman run --rm --entrypoint cat registry.ci.openshift.org/ci/ci-operator /usr/bin/ci-operator > ci-operator
$ file ci-operator
ci-operator: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, Go BuildID=0zvyBlXh4_7ZatNIZPjt/uu9NARUA2TtdjUy55Vpo/PHVfZ5sNLyxrFtYBPXhZ/-yq-meUB9qiZCfE3PpAH, not stripped
{{< / highlight >}}

For developers wanting to make changes to the code, the alternative is to build
it from source.  Development happens in the [`openshift/ci-tools`
repository][ci_tools].  It contains instructions on how to build it, but a
simple `go build ./cmd/ci-operator` should do it.

Once that is done, verify that you can execute the binary (it doesn't need to be
in `$PATH` as shown here):

{{< highlight bash >}}
$ ci-operator --help |& head -n 5
INFO[2022-03-09T16:21:18Z] unset version 0
Orchestrate multi-stage image-based builds

The ci-operator reads a declarative configuration YAML file and executes a set of build
steps on an OpenShift cluster for image-based components. By default, all steps are run,
{{< / highlight >}}

> The initial attempt to execute it will be frustrated, however:

Author: Changed to "without the appropriate setup", since there is more involved than arguments.

{{< highlight bash >}}
$ ci-operator
INFO[2022-03-09T16:23:09Z] unset version 0
ERRO[2022-03-09T16:23:09Z] Failed to load arguments.                     error=failed to determine job spec: no --git-ref passed and failed to resolve job spec from env: malformed $JOB_SPEC: $JOB_SPEC unset
{{< / highlight >}}

To understand why, a description of the usual environment in which `ci-operator`

> is executed is necessary.

Reviewer: Ack. I am good.

can be easily written manually.  For now, even a minimal definition suffices.
However, code in Prow and `ci-tools` generally assumes a well-formed definition,
so it is best to supply a valid value unless something very specific is begin
tested.  The following examples assume a definition similar to what the
`--git-ref` argument shown above generates:

{{< highlight bash >}}
export JOB_SPEC='{
  "type": "postsubmit",
  "refs": {
    "org": "openshift",
    "repo": "ci-tools",
    "base_ref": "master",
    "base_sha": "ef415037c6f8942784ee8a15635d5f0ce8a9cfcc"
  },
  "decoration_config": {
    "gcs_configuration": {
      "bucket": "origin-ci-test",
      "path_strategy": "single",
      "default_org": "openshift",
      "default_repo": "origin",
      "mediaTypes": {"log": "text/plain"}
    },
    "utility_images": {
      "clonerefs": "registry.ci.openshift.org/ci/clonerefs:latest",
      "initupload": "registry.ci.openshift.org/ci/initupload:latest",
      "entrypoint": "registry.ci.openshift.org/ci/entrypoint:latest",
      "sidecar": "registry.ci.openshift.org/ci/sidecar:latest"
    }
  }
}'
{{< / highlight >}}

> 

Author: Yes, this is used for all subsequent examples (and also happens to be the value I use for my local tests).

## `ci-operator-configresolver`

In a pattern that will soon become recurrent, attempting to execute
`ci-operator` with this new definition still results in an error, but a
different one:

{{< highlight bash >}}

> $ ci-operator

Author: No, this is the correct invocation here.  It is meant to show what happens when no arguments are passed.

The document is structured so that it shows the situations someone starting to test `ci-operator` may encounter and explains why they happen.

We will leave it be until a [later section]({{< relref "#test-namespaces" >}})
and focus on the other log line that appeared in the output.  There is now a
`JOB_SPEC`, so the <i>ref</i>s in it are used directly and simply echoed back in
a slightly different manner, but the second line is completely new.

One of the other major `ci-operator` input values is its configuration file.
Since no other information was given, it is derived from the repository
configuration that is now being provided.  This is done with assistance from the
`ci-operator-configresolver` service at `config.ci.openshift.org`, which has its
own [documentation page][configresolver].  For our purposes, it is enough to
mention that this is an easy way (both for CI jobs and for users) to obtain the
latest configuration file for a given input in a format that is directly usable
by `ci-operator`.  We will assume that is not what is intended and will instead
now detail the configuration file format.

Before leaving the resolver, we will note that it has other, secondary
responsibilities beyond serving configuration files.  It also serves
https://steps.ci.openshift.org and, in particular, a [complete

> reference][ci_operator_reference] of the fields of the configuration file.

Author: It [is](https://github.com/openshift/ci-tools/blob/c37de57c6d8a51ed73637e7fb8c9f5c0121added/pkg/webreg/zz_generated.ci_operator_reference.go), but it is ["served"](https://github.com/openshift/ci-tools/blob/c37de57c6d8a51ed73637e7fb8c9f5c0121added/pkg/webreg/webreg.go#L1650-L1654) to users by this program: https://steps.ci.openshift.org/ci-operator-reference.

> reference][ci_operator_reference] of the fields of the configuration file.

Reviewer: Got it.

definitions to guide the scheduling of test processes, which is done using the
(required) `resources` field.  It is a mapping field whose key is a name and
whose value is a definition of resources required by the test `Pod`, as
[specified by Kubernetes][kubernetes_resources].  A unique entry for the test
can be added under the `test` key (the name of the test), or a generic one can
be added using the special `*` key, which defines a default.

{{< highlight yaml >}}
# test.yaml
tests:
  # as before
resources:
  "*":
    requests:
      cpu: 100m
      memory: 8M # ought to be enough for anybody
{{< / highlight >}}

{{< highlight bash >}}
$ ci-operator --config test.yaml
INFO[2022-03-09T18:15:32Z] unset version 0
ERRO[2022-03-09T18:15:32Z] Failed to load arguments.                     error=tests[test].from: unknown image "quay.io/fedora/fedora"
{{< / highlight >}}

The image declaration (`from`) for the container test will not work as defined.
This is because, in order to determine the state of all the inputs listed in the
configuration file, `ci-operator` requires that all resources be local to the
cluster.  For this reason, direct references to external images (e.g. from Quay)

> such as the one we are attempting are not allowed.  The process for [configuring

Author: "We are attempting a direct reference" → "direct references [such as the one we are attempting] are not allowed" was the intent.

{{< highlight yaml >}}
base_images:
  fedora:
    namespace: ci
    name: fedora
    tag: latest
tests:
- as: test
  commands: echo test
  container:
    from: fedora
    clone: false
resources:
  # as before
{{< / highlight >}}

`clone: false` was added since by default tests which use one of the
`base_images` have the source code of the repository automatically cloned.  Our
configuration file is finally valid and ready to guide the execution of the test
it defines, but a few topics need to be covered before we can see the result.

## Test namespaces

Even though `ci-operator` is being executed directly on our machine, that is not
where most of the actual work is done.  Image builds, test executions, etc. must
happen inside an Openshift cluster, and `ci-operator` expects to be executed in
an environment where it has access to one.  This is why this final error message
mentions "in-cluster configuration" and missing Kuberentes host/port
information.  As mentioned before, the default environment is assumed to be a
Prow job, which is executed inside a Kubernetes `Pod`.  There, those environment
variables are provided automatically by Kubernetes (see the [service

> documentation][services_networking]).

Author: Hmm, that document must have changed since I wrote this.  This is meant to show the expected environment in actual jobs, so changed to one that explains those variables.

Outside of a Kubernetes cluster, the connection information can be passed using
the standard `KUBECONFIG` environment variable, which should point to a file

> just as `kubectl` and/or `oc` would expect it.  Note that the default context

Author: Didn't notice the ambiguity of taking "default" literally =)

  1. Via the command-line flag --kubeconfig
  2. Via the KUBECONFIG environment variable
  3. In your home directory as ~/.kube/config

To view or setup config directly use the 'config' command.
$ export KUBECONFIG=…
$ oc whoami --show-context
bbguimaraes/api-build01-ci-devcluster-openshift-com:6443/bbarcaro
{{< / highlight >}}

Our `ci-operator` is finally ready to execute its test, but one last adjustment
will make it a more pleasant experience.  In Kubernetes, isolation between
workloads is achieved using [_namespaces_][namespaces].  Openshift wraps
upstream namespaces into its own resource, the [_project_][project].  For most
purposes, the distinction is irrelevant and both names are used interchangeably.
`ci-operator` uses an ephemeral namespace to guarantee no interference between
unrelated tests, which is referred to as the _test namespace_.  The process by
which this namespace is set up will be discussed later; of relevance here is
that this setup is inconvenient for manual testing, for various reasons.

We can create our own test namespace in advance and simply ask `ci-operator` to

> use it instead of attempting to create one.  The build clusters are shared

Author: That's what I use (mostly because I haven't changed my `kubeconfig` from the days we only had `build01` =), but any cluster will do.

This is meant for Test Platform, so if someone doesn't have permission, they should be given.

> use it instead of attempting to create one.  The build clusters are shared

Reviewer: > This is meant for Test Platform

Gocha

the job begins.

The next line is the sequence of execution steps that will be performed.  It is
a comma-separated list of names, each of which corresponds to one task performed
by `ci-operator` for a given test (these are called "steps" internally and are
described in [this page][steps]).  Here, it shows the input image declared in
`base_images` and the test, which are discrete steps in the execution (an
[`InputImageTagStep`]({{< relref "steps#inputimagetagstep" >}}) and a
[`TestStep`]({{< relref "steps#teststep" >}}), respectively).

In reality, steps are organized into an execution _graph_, as described in the
[introduction][index], where some depend on others for their own execution.  In
the current example, the test cannot be scheduled before the image is imported:
the test is _dependent_ on the image import.  One useful property of the list in
the output is that it is a _topological ordering_ of the graph, i.e. dependents
always come before the steps they depend on.  As a one-dimensional projection of
a partial order, it is of limited usefulness in more complicated graphs with
multiple dependencies between nodes, but it gives a general idea of the order of
execution.

> The next two lines are each output by their respective step when it is executed.

Reviewer: ```suggestion
Each of the next two lines is the output of their respective step when it is executed.
```

The image import describes its input and output (the meaning of `pipeline` will
be explained in a [later section]({{< relref "#pipeline" >}})) and the test

> simply notifies that it is being executed.  Finally, the total execution time is

Reviewer: "Executing test test": Why do we have "test" twice there?

Author: The message is "Executing test ${test_name}".  I use `%q` in log messages exactly for this reason, but this one comes from the lizard brain of `ci-operator`.

displayed right at the end of the execution of `ci-operator`.

Why was the `--target` argument necessary?  Remember that `ci-operator`'s
defaults are optimized for its use in CI jobs.  As such, it implicitly includes

> common job tasks in the execution graph:

Reviewer: "common job tasks": What are they?

I thought it would list _all_ steps in the graph without `--target`. No?

Author: The extra ones in the `Running` line below: `[output-images]`, `[images]`, `src`.  They are not exactly "all steps", since there are others we choose not to include.  Building the `src` image in this case, for example, is an arbitrary decision since it's "commonly" done.  This is mostly a curiosity, since all jobs have a `--target` argument.

Reviewer: :TIL
